File: frontend/lib/models/grade.dart
class Grade {
  final int id;
  final String subject;
  final String value;
  final DateTime createdAt;
  final String teacherName;
  final String? classroomName;
  final String? studentName;

  Grade({
    required this.id,
    required this.subject,
    required this.value,
    required this.createdAt,
    required this.teacherName,
    this.classroomName,
    this.studentName,
  });

  factory Grade.fromJson(Map<String, dynamic> json) {
    return Grade(
      id: json['id'] as int,
      subject: json['subject'] as String,
      value: json['value'] as String,
      createdAt: DateTime.parse(json['created_at'] as String),
      teacherName: json['teacher_name'] as String,
      classroomName: json['classroom_name'] as String?,
      studentName: json['student_name'] as String?,
    );
  }
}

File: backend/app/schemas/grade.py
from datetime import datetime

from pydantic import BaseModel, Field, field_validator


class GradeCreate(BaseModel):
    student_id: int
    classroom_id: int | None = None
    subject: str = Field(min_length=1, max_length=100)
    value: str = Field(min_length=1, max_length=20)

    @field_validator("subject", "value")
    @classmethod
    def not_blank(cls, v: str) -> str:
        if not v.strip():
            raise ValueError("must not be blank")
        return v.strip()


class GradeOut(BaseModel):
    id: int
    subject: str
    value: str
    created_at: datetime
    teacher_name: str
    classroom_name: str | None = None
    student_name: str | None = None
